src/transferslot.cpp

```diff
@@ -347,7 +347,100 @@ int64_t TransferSlot::macsmac(chunkmac_map* m)
     return m->macsmac(transfer->transfercipher());
 }
 
-bool TransferSlot::checkTransferFinished(DBTableTransactionCommitter& committer, MegaClient* client)
+int64_t TransferSlot::macsmac_gaps(chunkmac_map* m, size_t g1, size_t g2, size_t g3, size_t g4)
+{
+    return m->macsmac_gaps(transfer->transfercipher(), g1, g2, g3, g4);
+}
+
+bool TransferSlot::checkMetaMacWithMissingLateEntries()
+{
+    // Due to an old bug, some uploads attached a MAC to the node that was missing some MAC entries 
+    // (even though the data was uploaded) - this occurred when a ultoken arrived but one other 
+    // final upload connection had not completed at the local end (even though it must have 
+    // completed at the server end).  So the file's data is still complete in the cloud.
+    // Here we check if the MAC is one of those with a missing entry (or a few if the connection had multiple chunks)
+
+    // last 3 connections, up to 32MB (ie chunks) each, up to two completing after the one that delivered the ultoken
+    size_t end = transfer->chunkmacs.size();
+    size_t finalN = std::min<int>(32 * 3, end);
```

Comment:
I guess that `3` comes from the fact that the maximum number of connections per transfer is 6, so there will be, at maximum, up to 3 gaps / missing entries. If that's the case, rather than `3`:
```suggestion
    size_t finalN = std::min<int>(32 * MegaClient::MAX_NUM_CONNECTIONS/2, end);
```